src/solver.c

```diff
@@ -694,12 +694,19 @@ static void select_package(struct apk_solver_state *ss, struct apk_name *name)
 			    p->pkg->name->ss.requirers == 0 &&
 			    (p->pkg->provider_priority == 0 && name->providers->num > 1))
 				continue;
+      /* check if package is ignored and already installed */
+      if ((p->pkg->ss.solver_flags & APK_SOLVERF_IGNORE_UPGRADE) && p->pkg->ipkg != NULL)
```

Comment:
I'd rather not keep this state here in `select_package()`. Please make `compare_providers` instead check for the new flag. That function already has a spot for preferring already installed package on self-upgrade, just make that check also use the new flag. Though, similar check might need to be added to the "latest" code path as well.

Reply:
This makes sense, I will move this flag check into `compare_providers`.